File: go/webpush/send_notification_with_action.go
package main

// More info about device_id:
//    (web): https://pushe.co/docs/webpush/#unique_id

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
)

func main() {

	// Obtain token -> https://pushe.co/docs/api/#api_get_token
	const token = "YOUR_TOKEN"

	// Webpush -> https://pushe.co/docs/webpush-api/

	reqData := map[string]interface{}{
		"app_ids":  []string{"YOUR_APP_ID"},
		"platform": 2,
		"data": map[string]interface{}{
			"title":   "This is a simple push",
			"content": "All of your users will see me",

			// Actions -> https://pushe.co/docs/webpush-api/#webpush_api_action_type_table2
			"action": map[string]interface{}{
				"action_type": "U",
				"url":         "https://pushe.co",
			},

			"buttons": []map[string]interface{}{
				{
					"btn_content": "Pushe",
					"btn_action": map[string]interface{}{
						"action_type":         "U",
						"url":                 "https://pushe.co",
					},
					"btn_order": 0,
				},
			},
		},
		// additional keywords -> https://pushe.co/docs/webpush-api/#api_send_advance_notification
	}

	// Marshal returns the JSON encoding of reqData.
	reqJSON, err := json.Marshal(reqData)

	// check encoded json
	if err != nil {
		fmt.Println("json:", err)
		return
	}

	// create request obj
	request, err := http.NewRequest(
		http.MethodPost,
		"https://api.pushe.co/v2/messaging/notifications/",
		bytes.NewBuffer(reqJSON),
	)

	// check request
	if err != nil {
		fmt.Println("Req error:", err)
		return
	}

	// set header
	request.Header.Set("Content-Type", "application/json")
	request.Header.Set("Accept", "application/json")
	request.Header.Set("Authorization", "Token "+token)

	// send request and get response
	client := http.Client{}
	response, err := client.Do(request)

	// check response
	if err != nil {
		fmt.Println("Resp error:", err)
		return
	}

	defer response.Body.Close()

	// check status_code and response
	buf := new(bytes.Buffer)
	_, _ = buf.ReadFrom(response.Body)
	respContent := buf.String()

	fmt.Println("status code =>", response.StatusCode)
	fmt.Println("response =>", respContent)
	fmt.Println("==========")

	if response.StatusCode == http.StatusCreated {
		fmt.Println("success!")

		var respData map[string]interface{}
		_ = json.Unmarshal(buf.Bytes(), &respData)

		var reportURL string

		// hashed_id just generated for Non-Free plan
		if respData["hashed_id"] != nil {
			reportURL = "https://pushe.co/report?id=" + respData["hashed_id"].(string)
		} else {
			reportURL = "no report url for your plan"
		}

		fmt.Println("report_url:", reportURL)
		fmt.Println("notification id:", int(respData["wrapper_id"].(float64)))
	} else {
		fmt.Println("failed!")
	}
}

File: python/webpush/send_notification_with_action.py
#!/usr/bin/env python
# vim: ts=4 sw=4 et

import requests

# obtain token -> https://pushe.co/docs/api/#api_get_token
TOKEN = 'YOUR_TOKEN'

# set header
headers = {
    'Authorization': 'Token ' + TOKEN,
    'Content-Type': 'application/json'
}

# Webpush -> https://pushe.co/docs/webpush-api/

data = {
    'app_ids': ['YOUR_APP_ID', ],
    'platform': 2,
    'data': {
        'title': 'this is the title',
        'content': 'this is the content',
    },

    # Actions -> https://pushe.co/docs/webpush-api/#webpush_api_action_type_table2
    'action': {
        'action_type': 'U',
        'url': 'https://pushe.co',
    },

    'buttons': [
        {
            'btn_content': 'YOUR_CONTENT',
            'btn_action': {
                'action_data': 'ActivityName',
                'action_type': 'T',
                'market_package_name': '',
                'url': ''
            },
            'btn_order': 0,
        }
    ],
}

# send request
response = requests.post(
    'https://api.pushe.co/v2/messaging/notifications/',
    json=data,
    headers=headers,
)

# get status_code and response
print('status code => ', response.status_code)
print('response => ', response.json())
print('==========')

if response.status_code == 201:
    print('Success!')

    data = response.json()

    # hashed_id only generated for Non-Free plan
    if data['hashed_id']:
        report_url = 'https://pushe.co/report?id=%s' % data['hashed_id']
    else:
        report_url = 'no report url for your plan'

    notif_id = data['wrapper_id']
    print('report_url: %s' % report_url)
    print('notification id: %s' % notif_id)
else:
    print('failed')
